core/src/price_estimation/clients/apiclient.rs

```diff
@@ -134,31 +143,71 @@ where
 
 #[cfg(test)]
 mod tests {
-    use super::api::MockDexagApi;
     use super::*;
+    use crate::http::HttpFactory;
+    use crate::models::TokenId;
     use crate::price_estimation::data::TokenBaseInfo;
-    use crate::util::FutureWaitExt as _;
+    use anyhow::{anyhow, Result};
     use lazy_static::lazy_static;
-    use mockall::{predicate::*, Sequence};
+    use mockall::{mock, predicate::*, Sequence};
+
+    #[derive(Clone, Debug, PartialEq)]
+    // Cannot autogenerate with Mockall since the derived traits are needed
+    // for testing. Symbolic is a trait that is assigned to the internal token
+    // representation of the price source, so the output of `.symbol()` isn't
+    // expected to change.
+    pub struct MockSymbolic(String);
+    impl Symbolic for MockSymbolic {
+        fn symbol(&self) -> &str {
+            &self.0
+        }
+    }
+    impl MockSymbolic {
+        fn create(name: &str) -> MockSymbolic {
+            MockSymbolic(name.to_string())
+        }
+    }
+
+    // Generate MockHttpApi, which mocks the two traits Api and HttpConnecting
+    // simultaneously as per Mockall documentation:
+    // https://docs.rs/mockall/0.5.1/mockall/macro.mock.html
+    // Note that trait definitions are not just a copypaste of the originals, since
+    // 1. referencing the associated type in the macro is forbidden, see docs above
+    // 2. must explicitly specify associated type in definition
+    // To get copypaste working, replace `Self::Token` -> `MockSymbolic` and
+    // `type Token` -> `type Token = MockSymbolic`
+    mock! {
+        pub HttpApi {}
+        pub trait Api {
+            type Token = MockSymbolic;
+            fn get_token_list<'a>(&'a self) -> BoxFuture<'a, Result<Vec<MockSymbolic>>>;
+            fn get_price<'a>(&'a self, from: &MockSymbolic, to: &MockSymbolic) -> BoxFuture<'a, Result<f64>>;
+        }
+        pub trait HttpConnecting: Sized {
+            fn bind(http_factory: &HttpFactory) -> Result<Self>;
+        }
+    }
```

Comment:
I think we don't need all this extra setup if we use the original token data type and instantiate our dummy test objects as such structs. This should make this change less intrusive.

Reply:
I don't like much the idea of having the test for a generic client implementation use a token defined for a specific API implementation. On the other hand, merging `Api` and `HttpConnecting` means that there is no need for the `mock!` macro anymore, so maybe the new changes are already enough to address this comment?

Reply:
Ah I didn't see before, but the Token struct will be different for the 1Inch API? That's interesting but I still think you don't necessarily need a trait for that and it would be much simpler for readability if we used a concrete token type in the tests instead (see comment above about passive data objects).

Reply:
Yeah, the token struct will be different for every API implementation. That's a key change I need to make to include 1inch, and likely other APIs. The reason for `Symbolic` is that I still need a way to convert the custom token structure to a more "standardized" token structure that is the same for different API. The conversion is done by `.symbol()`, because the only thing we really need of a token to implement `PriceSource` is its symbol. (That's also nice because it allows you to handle weird conversions when implementing `Symbolic`. For example, some APIs use "ETH" to mean "WETH", so this conversion would be done cleanly in `.symbol()`.)
So, whithin these limitations the extra `Symbolic` trait seemed a good option.
For the tests, I could rename `MockSymbolic` to `MockToken` to clarify the purpose of it (i could even rename `Symbolic` to `Tokenized`, `Tokenizing` or something like that), but I don't see how I could simplify the implementation of `MockSymbolic` in any other way.

Reply:
My main point is that when you just need something that carries data, instead of going through all the code required for the trait, we could just make a struct that contains the data and pass this around being. 
In this very example you even only have one field (a string), so why not just pass symbol as a string directly?